### PR TITLE
Allowed instructors

### PR DESCRIPTION
**Pull request by Kingofikings441**
When predefining a course, can only assign eligible teachers. The form also sorts by preference then bandwidth(how many courses they can still take)

### CODE REVIEW

File: app/controllers/room_bookings_controller.rb [hunk 1]
```diff
@@ -128,16 +132,34 @@ def toggle_lock
...
+    # @allowed_instructors = eligible_instructors(@schedule, time_slot).sort_by do |instructor|
+    #   # You can customize this sorting logic based on the structure of your `InstructorPreference` model.
+    #   # For example, if `instructor.instructor_preferences` has a `preference_level` attribute,
+    #   # you could sort by this value:
+    #   preference = instructor.instructor_preferences.find_by(course_id: @course.id)
+    #   preference_level = preference.preference_level || 0
+  
+    #   # Return preference_level for sorting, instructors with lower preference level will come first
+    #   preference_level
+    # end.reverse
```
Comment (Naviunni) [inline app/controllers/room_bookings_controller.rb]: Please remove these comments

File: app/controllers/room_bookings_controller.rb [hunk 2]
```diff
@@ -128,16 +132,34 @@ def toggle_lock
   end
 
   def update_instructor
+    @schedule = Schedule.find(params[:schedule_id])
```
Comment (Naviunni) [inline app/controllers/room_bookings_controller.rb:135]: create a private set_schedule function similar to other controllers

File: app/controllers/room_bookings_controller.rb [hunk 3]
```diff
@@ -128,16 +132,34 @@ def toggle_lock
...
+    
     if @booking.update(instructor_id: params[:room_booking][:instructor_id])
       flash[:notice] = 'Instructor updated successfully.'
     else
       flash[:alert] = 'Failed to update instructor.'
     end
 
     # Redirect to the previous page or another relevant page
-    redirect_back(fallback_location: room_bookings_path)
+    # redirect_back(fallback_location: room_bookings_path)
```
Comment (Naviunni) [inline app/controllers/room_bookings_controller.rb]: remove comment

File: app/controllers/room_bookings_controller.rb [hunk 4]
```diff
@@ -275,4 +297,33 @@ def section_alloted?(overlapping_booking)
...
+    
+  # end
+
+  # def get_teach_count(instructor)
+  #   bookings_for_instructor = RoomBooking.all.select do |booking|
+  #     booking.instructor_id == instructor.id
+  #   end
+  #   bookings_for_instructor.length
+  # end
+
```
Comment (Naviunni) [inline app/controllers/room_bookings_controller.rb:291]: Same, please remove these comments

**Review by Naviunni — CHANGES_REQUESTED** [review]
Please add some cucumber tests as well

File: app/controllers/room_bookings_controller.rb (hunk 4 above)
Comment (Kingofikings441) [reply app/controllers/room_bookings_controller.rb:291]: removed

File: app/controllers/room_bookings_controller.rb (hunk 3 above)
Comment (Kingofikings441) [reply app/controllers/room_bookings_controller.rb]: removed

File: app/controllers/room_bookings_controller.rb (hunk 2 above)
Comment (Kingofikings441) [reply app/controllers/room_bookings_controller.rb:135]: created

File: app/controllers/room_bookings_controller.rb (hunk 1 above)
Comment (Kingofikings441) [reply app/controllers/room_bookings_controller.rb]: removed

File: app/controllers/room_bookings_controller.rb [hunk 5]
```diff
@@ -128,16 +132,24 @@ def toggle_lock
   end
 
   def update_instructor
+    @schedule = Schedule.find(params[:schedule_id])
     @booking = RoomBooking.find(params[:id])
-
+    @section = Section.find_by(id: @booking.section_id)
```
Comment (wahibkapdi) [inline app/controllers/room_bookings_controller.rb:137]: Sections are no longer part of our flow.

Comment (wahibkapdi) [reply app/controllers/room_bookings_controller.rb:137]: RoomBookings are directly linked to courses now.

File: app/helpers/room_bookings_helper.rb [hunk 6]
```diff
@@ -9,4 +9,94 @@ def available_sections(schedule)
                        .where(room_booking: { id: nil })
     render partial: '/shared/courses_list', locals: { sections: @sections }
   end
+  def eligible_instructors(schedule, booking)
+    time_slot = TimeSlot.find_by(id: booking.time_slot_id)
+    current_instructor_id = booking.instructor_id 
+    course_id = booking.section.course.id
```
Comment (wahibkapdi) [inline app/helpers/room_bookings_helper.rb:15]: Same as above.

File: app/controllers/room_bookings_controller.rb [hunk 7]
```diff
@@ -275,4 +286,7 @@ def section_alloted?(overlapping_booking)
   def room_booking_params
     params.require(:room_booking).permit(:room_id, :time_slot_id, :is_available, :is_lab, :instructor_id, :section_id)
```
Comment (wahibkapdi) [inline app/controllers/room_bookings_controller.rb:287]: Same as above.

File: app/views/room_bookings/index.html.erb [hunk 8]
```diff
@@ -56,6 +56,12 @@
                                     (<%= booking.section.section_number %>)
                                 </strong> |
                                 <%= booking.section.seats_alloted %>
```
Comment (wahibkapdi) [inline app/views/room_bookings/index.html.erb:58]: Same as above.

File: app/views/room_bookings/index.html.erb [hunk 9]
```diff
@@ -74,18 +80,24 @@
...
+                                    <%= form_with model: booking, 
+                                        url: update_instructor_schedule_room_booking_path(@schedule, booking), 
+                                        method: :patch,
+                                        remote: true, 
+                                        data: { turbo_method: :patch } do |form| %>
+                                        <div class="field">
+                                            <%= form.select :instructor_id, 
+                                                # options_from_collection_for_select(eligible_instructors(@schedule,  booking)  , :id, :first_name, booking.instructor_id), 
+                                                options_for_select(
+                                                    eligible_instructors(@schedule, booking).map { |instructor| ["#{instructor.first_name} #{instructor.last_name} #{get_preference(instructor, booking.section.course.id)}", instructor.id] },
```
Comment (wahibkapdi) [inline app/views/room_bookings/index.html.erb:93]: Same as above.

**Review by wahibkapdi — CHANGES_REQUESTED** [review]

**Comment by wahibkapdi** [comment]
Closing as another PR with the required changes has been merged.

**Pull request closed without merging**